smu/geometry/topology_from_geom.py

```diff
@@ -0,0 +1,174 @@
+        shortest_distance = distances[a1, a2]
+        closest_heavy_atom = a2
+
+    if closest_heavy_atom < 0:
+      return None
+
+    bond = dataset_pb2.BondTopology.Bond(atom_a=a1,
+                                         atom_b=closest_heavy_atom,
+                                         bond_type=dataset_pb2.BondTopology.BondType.BOND_SINGLE)
+    result.bonds.append(bond)
+
+  return result
+
+
+def indices_of_heavy_atoms(bond_topology: dataset_pb2.BondTopology) -> List[int]:
+  """Return the indices of the heavy atoms in `bond_topology`.
+    Args:
+    Returns:
+  """
+  return [
+      i for i, t in enumerate(bond_topology.atoms) if t != dataset_pb2.BondTopology.AtomType.ATOM_H
+  ]
+
+
+def bond_topologies_from_geom(
+    bond_lengths: bond_length_distribution.AllAtomPairLengthDistributions,
+    bond_topology: dataset_pb2.BondTopology, geometry: dataset_pb2.Geometry,
+    matching_parameters: smu_molecule.MatchingParameters) -> dataset_pb2.TopologyMatches:
+  """Return all BondTopology's that are plausible.
+
+    Given a molecule described by `bond_topology` and `geometry`, return all possible
+    BondTopology that are consistent with that.
+    Note that `bond_topology` will be put in a canonical form.
+    Args:
+      bond_length_distribution:
+      bond_topology:
+      geometry:
+    Returns:
+      TopologyMatches
+  """
+  result = dataset_pb2.TopologyMatches()    # To be returned.
+  if len(bond_topology.atoms) == 1:
+    return result    # empty.
+
+  # Will be used when comparing perceived BondTopology's.
+  serialized_starting_form = bond_topology.SerializeToString()
+
+  utilities.canonical_bond_topology(bond_topology)
+  distances = utilities.distances(geometry)
+
+  # First join each Hydrogen to its nearest heavy atom, thereby
+  # creating a starting BondTopology from which all others can grow
+  starting_bond_topology = hydrogen_to_nearest_atom(bond_topology, distances)
+  if starting_bond_topology is None:
+    return result
+
+# heavy_atoms = [a for a in bond_topology.atoms if a != dataset_pb2.BondTopology.AtomType.ATOM_H]
+# heavy_atom_indices = indices_of_heavy_atoms(bond_topology)
+  heavy_atom_indices = [
+      i for i, t in enumerate(bond_topology.atoms) if t != dataset_pb2.BondTopology.AtomType.ATOM_H
+  ]
+
+  # For each atom pair, a list of possible bond types.
+  # Key is a tuple of the two atom numbers, value is an np.array
+  # with the score for each bond type.
+
+  bonds_to_scores: Dict[Tuple[int, int], np.array] = {}
+  for c in itertools.combinations(heavy_atom_indices, 2):    # All pairs.
+    i = c[0]
+    j = c[1]
+    dist = distances[i, j]
+    if dist > THRESHOLD:
+      continue
+    btypes = np.zeros(4, np.float32)
+    for btype in range(0, 4):
+      #    print(f"Looking for pdfs of {bond_topology.atoms[i]} {bond_topology.atoms[j]} type {btype} dist {dist}")
+      btypes[btype] = bond_lengths.pdf_length_given_type(bond_topology.atoms[i],
```

Comment:
This is the wrong function to use. 
You should be using 
bond_lengths.probability_of_bond_types
because that gives you real probabilities and not pdfs. 
That also only returns the plausible bond types but you can move those back to zero if you need.

Reply:
THis is an interesting one. 
Think about the case of a pair of atoms that are separated b a large distance - the longest ever single bond observed. That pair would have a 100% probability of being classified as a single bond - even though it is a very unusual single bond.
Using independent scores on the other hand lets the overall score reflect this poor fit.
When comparing two possible BT's for a given geometry, using a normalized probability would favor the poor matches - since the probability of a single (or triple) bond would be very high.
This is an important point, let's talk more....